### PR TITLE
[Config] type specific check for emptiness

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13736 |
| License | MIT |
| Doc PR |  |

### CODE REVIEW

**Comment by WouterJ** [comment]
:+1:

**Comment by xabbuh** [comment]
ping @symfony/deciders

**Comment by fabpot** [comment]
:+1:

**Comment by dunglas** [comment]
:+1:

**Comment by Tobion** [comment]
I'd rather say, as also stated in the ticket, that `cannotBeEmpty` on a boolean node (etc.) makes no sense. So it should rather throw an exception.

So I'm +-0 with this solution.

**Comment by xabbuh** [comment]
I thought about that too, but decided to go this way as throwing an exception would potentially break many existing applications.

**Comment by Tobion** [comment]
Then just change it in 2.8 or 3.0

**Comment by xabbuh** [comment]
Then I suggest to mark the method as deprecated in 2.8 and throw the exception in 3.0. Do you agree with this plan?

**Comment by fabpot** [comment]
@xabbuh I agree

**Comment by xabbuh** [comment]
see #15285 for the deprecations

**Comment by xabbuh** [comment]
I just discovered that the changes did not cover float nodes properly. I moved the check from `IntegerNode` to `NumericNode`.

**Comment by fabpot** [comment]
Thank you @xabbuh.

File: src/Symfony/Component/Config/Tests/Definition/ScalarNodeTest.php [hunk 1]
```diff
@@ -57,4 +57,51 @@ public function getInvalidValues()
...
+    {
+        $node = new ScalarNode('test');
+        $node->setAllowEmptyValue(false);
+
+        $this->assertSame($value, $node->finalize($value));
+    }
+
+    public function getValidNonEmptyValues()
+    {
+        return array(
```
Comment (SoboLAN) [inline src/Symfony/Component/Config/Tests/Definition/ScalarNodeTest.php]: What about `"0"` ?

Comment (xabbuh) [inline src/Symfony/Component/Config/Tests/Definition/ScalarNodeTest.php]: What do you mean?

Comment (SoboLAN) [inline src/Symfony/Component/Config/Tests/Definition/ScalarNodeTest.php]: Well shouldn't that value be there as well ?

Maybe I'm wrong, I just wanted to make sure you didn't accidentally miss it :)

Comment (xabbuh) [inline src/Symfony/Component/Config/Tests/Definition/ScalarNodeTest.php]: It's true that we could have added it, but I think it's not absolutely necessary. Of course, if you think it deserves it own test, please don't hesitate to make a pull request for it. :)

**Pull request merged**